### PR TITLE
Adding more Rbac standard user tests (43 and 44) with specifc verbs

### PR DESCRIPTION
**Pull request by mmartin24**
Implementation of [FLEET-43](https://app.qase.io/project/FLEET?previewMode=modal&suite=13&tab=&case=43) and [FLEET-44](https://app.qase.io/project/FLEET?previewMode=side&suite=13&tab=&case=44)

#### Additional implementations:
- Refactor of `createRoleTemplate` to add specific verbs per resource
- Further refactor of Fleet-5 to be compliant with previous refactor
- Other refactors to ensure test reliability

### CODE REVIEW

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts [hunk 1]
```diff
@@ -38,9 +38,11 @@ describe('Test Fleet access with RBAC with custom roles', { tags: '@rbac' }, ()
       cy.createRoleTemplate({
         roleType: "Global",
         roleName: "fleetAllVerbsRole",
-        newUserDefault: "Yes",
-        resources: ["fleetworkspaces", "gitrepos", "bundles"],
-        verbs: ["create", "delete", "get", "list", "patch", "update", "watch"],
+        resources: [
```
Comment (thehejik) [inline tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:41]: IMO `resources` is not valid name anymore for this array, it should be called `rules` instead. Each entry/object is `rule`.

File: tests/cypress/support/commands.ts [hunk 2]
```diff
@@ -262,22 +262,23 @@ Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUse
...
-        verbs.forEach((verb) => {
-          cy.get(`input.vs__search`).eq(2 * i).click();
-          cy.get(`ul.vs__dropdown-menu > li`).contains(verb).should("exist").click();
-        });
-      }
-    });
-  }
+    // Addition of resources and verbs linked to resources
+    // Each resource is an object with 2 keys: resource and verbs
+    if (resources) {
```
Comment (thehejik) [inline tests/cypress/support/commands.ts:267]: I would remove these `if`s - if it is empty array the `forEach`  won't iterate over it.

File: tests/cypress/support/commands.ts [hunk 3]
```diff
@@ -262,22 +262,23 @@ Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUse
...
+      resources.forEach((resource: { resource: string, verbs: string[] }, i) => {
+        // Iterate over Resource cells and add 1 resource
+        cy.get(`input.vs__search`).eq(2 * i + 1).click();
+        cy.contains(resource.resource, { matchCase: false }).should("exist").click();
+        cy.clickButton("Add Resource");
+  
+        if (resource.verbs) {
```
Comment (thehejik) [inline tests/cypress/support/commands.ts:274]: same

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts (hunk 1 above)
Comment (sbulage) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:41]: `resourcesWithRules` would be appropriate name.

Comment (thehejik) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:41]: > `resourcesWithRules` or `resourcesWithVerbs` would be appropriate name.

I disagree, the corresponding block in ClusterRole resource is called `rules` in k8s, please see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-examples

Comment (mmartin24) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts]: Comment address in commit `34f79df`

File: tests/cypress/support/commands.ts (hunk 3 above)
Comment (mmartin24) [reply tests/cypress/support/commands.ts]: Comment addressed in commit `34f79df`

File: tests/cypress/support/commands.ts [hunk 4]
```diff
@@ -262,22 +262,19 @@ Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUse
     cy.get('span[aria-label="Yes: Default role for new users"]').click();
   }
   
-  // Addition of resources and verbs linked to resources
-  if (resources) {
-    resources.forEach((resource, i) => {
+    // Addition of resources and verbs linked to resources
+    // Each resource is an object with 2 keys: resource and verbs
+    rules.forEach((resource: { resource: string, verbs: string[] }, i) => {
```
Comment (thehejik) [inline tests/cypress/support/commands.ts:267]:
```suggestion
    rules.forEach((rule: { resource: string, verbs: string[] }, i) => {
```

File: tests/cypress/support/commands.ts [hunk 5]
```diff
@@ -262,22 +262,19 @@ Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUse
...
       // Iterate over Resource cells and add 1 resource
       cy.get(`input.vs__search`).eq(2 * i + 1).click();
-      cy.contains(resource, { matchCase: false }).should("exist").click();
+      cy.contains(resource.resource, { matchCase: false }).should("exist").click();
```
Comment (thehejik) [inline tests/cypress/support/commands.ts:270]:
```suggestion
      cy.contains(rule.resource, { matchCase: false }).should("exist").click();
```

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts [hunk 6]
```diff
@@ -62,4 +65,90 @@ describe('Test Fleet access with RBAC with custom roles', { tags: '@rbac' }, ()
...
+      cy.wait(1000)
+      cy.accesMenuSelection('Continuous Delivery', 'Advanced', 'Bundles');
+      cy.accesMenuSelection('Continuous Delivery', 'Advanced', 'Workspace');
+      cy.get('a.btn.role-primary').contains('Create').should('be.visible');
+      cy.verifyTableRow(0, 'Active', 'fleet-default');
+      cy.verifyTableRow(1, 'Active', 'fleet-local');
+      cy.open3dotsMenu('fleet-default', 'Edit Config');
+      cy.contains('allowedTargetNamespaces').should('be.visible');
+      
+      // What the user should NOT be able to access (once in Fleet Workspace menu)
```
Comment (sbulage) [inline tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:148]: Comment is unclear :thinking:

File: tests/cypress/support/commands.ts [hunk 7]
```diff
@@ -262,22 +262,19 @@ Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUse
...
       cy.clickButton("Add Resource");
 
-      if (verbs) {
-        verbs.forEach((verb) => {
+        resource.verbs.forEach((verb) => {
```
Comment (thehejik) [inline tests/cypress/support/commands.ts:273]:
```suggestion
        rule.verbs.forEach((verb) => {
```

File: tests/cypress/support/e2e.ts [hunk 8]
```diff
@@ -35,7 +35,7 @@ declare global {
       checkApplicationStatus(appName: string, clusterName?: string): Chainable<Element>;
       deleteApplicationDeployment(clusterName?: string): Chainable<Element>;
       modifyDeployedApplication(appName: string, clusterName?: string): Chainable<Element>;
-      createRoleTemplate(roleType: string, roleName: string, newUserDefault?: string['yes'|'no'], verbs?: string[], resources?: string[], apiGroups?: string[], nonResourcesURLs?: string[] ): Chainable<Element>;
+      createRoleTemplate(roleType: string, roleName: string, newUserDefault?: string['yes'|'no'], verbs?: string[], rules?: string[], apiGroups?: string[], nonResourcesURLs?: string[] ): Chainable<Element>;
```
Comment (thehejik) [inline tests/cypress/support/e2e.ts:38]: Here not sure whether it should contain `verbs` which is part of `rules`.

```suggestion
      createRoleTemplate(roleType: string, roleName: string, newUserDefault?: string['yes'|'no'], rules?: string[], apiGroups?: string[], nonResourcesURLs?: string[] ): Chainable<Element>;
```

File: tests/cypress/support/commands.ts [hunk 9]
```diff
@@ -247,7 +247,7 @@ Cypress.Commands.add('modifyDeployedApplication', (appName, clusterName='local')
 });
 
 // Create Role Template (User & Authentication)
-Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUserDefault='No', verbs, resources, apiGroups, nonResourcesURLs}) => {
+Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUserDefault='No', verbs, rules, apiGroups, nonResourcesURLs}) => {
```
Comment (thehejik) [inline tests/cypress/support/commands.ts:250]: Verbs are part of rules, so not sure if it should be here

```suggestion
Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUserDefault='No', rules, apiGroups, nonResourcesURLs}) => {
```

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts [hunk 10]
```diff
@@ -62,4 +65,90 @@ describe('Test Fleet access with RBAC with custom roles', { tags: '@rbac' }, ()
     })
   )
 
+  qase(43,
+    it('Test "Std-Base" user with custom role "list" and "create" to "fleetworkspaces" can list and create workspaces but NOT  "edit" nor "delete" them', { tags: '@fleet-43' }, () => {
```
Comment (sbulage) [inline tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:69]: Please use full name i.e. 
`'Test 'Standard Base' Role user with "list" and "create" verbs for "fleetworkspaces" resource. User can NOT "edit" or "delete" them'`

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts [hunk 11]
```diff
@@ -62,4 +65,90 @@ describe('Test Fleet access with RBAC with custom roles', { tags: '@rbac' }, ()
...
+      // What the user should be able to access
+      cy.accesMenuSelection('Continuous Delivery', 'Git Repos');
+      cy.wait(500)
+      cy.accesMenuSelection('Continuous Delivery', 'Advanced', 'Bundles');
+      cy.accesMenuSelection('Continuous Delivery', 'Advanced', 'Workspaces');
+      cy.verifyTableRow(0, 'Active', 'fleet-default');
+      cy.verifyTableRow(1, 'Active', 'fleet-local');
+      cy.get('a.btn.role-primary').contains('Create').should('be.visible');
+      
+      // What the user should NOT be able to access
```
Comment (sbulage) [inline tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:104]: // What `verbs` the user should NOT be able to access <-- Please add highlighted word for better understanding.

**Review by sbulage — COMMENTED** [review]
Not a blocker but good to have. LGTM :+1:

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts (hunk 11 above)
Comment (sbulage) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:104]: This applies to all places where this comment it used.

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts (hunk 6 above)
Comment (mmartin24) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:148]: What is unclear?
It points to what the user should not be able to access

**Comment by thehejik** [comment]
BTW `apiGroups` and `nonResourcesUrls` could be also part of the rules array. This can be done later.

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts (hunk 6 above)
Comment (sbulage) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:148]: `verbs` word is not in that comment..

Comment (sbulage) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:148]: Like: // What verb(s) user should NOT be able to access (once in Fleet Workspace menu)

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts (hunk 11 above)
Comment (mmartin24) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:104]: Verbs not to to be able to access are specified on the title of the test.
I added this comment for clearness where to focus. I could remove it.
I think is redundant to point again the verbs mentioned in title, but I will if desired

Comment (sbulage) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:104]: Please, don't remove it. They are good. I know test is using verbs but here clearly mentioning about the access to the specified menu by segregating with verbs would be much clearer, IMO :smile: 
Feel free to keep it as it is no worries :smile:

Comment (mmartin24) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:104]: Changed to:
```
// Ensuring the user is not able to "edit" or "delete" workspaces.
```

Comment (sbulage) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:104]: Fine with that Thank You :smile:

File: tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts (hunk 6 above)
Comment (mmartin24) [reply tests/cypress/e2e/unit_tests/rbac_fleet.spec.ts:148]: I thought it was clear enough by the title pointing that should not be able to delete and by next command:
```
cy.open3dotsMenu('fleet-default', 'Delete', true);
```

But ok will change to this:

```
// Ensuring the user is not able to "delete" workspaces. 
```

File: tests/cypress/support/commands.ts [hunk 12]
```diff
@@ -258,26 +258,23 @@ Cypress.Commands.add('createRoleTemplate', ({roleType='Global', roleName, newUse
...
-  if (newUserDefault === 'Yes') {
+  if (newUserDefault === 'yes') {
     cy.get('span[aria-label="Yes: Default role for new users"]').click();
   }
   
-  // Addition of resources and verbs linked to resources
-  if (resources) {
-    resources.forEach((resource, i) => {
+    // Addition of resources and verbs linked to resources
+    // Each resource is an object with 2 keys: resource and verbs
```
[the remark this answers is not in the document]
Comment (mmartin24) [reply tests/cypress/support/commands.ts:266]: ok. Removed as well

[the remark this answers is not in the document]
Comment (mmartin24) [reply tests/cypress/support/commands.ts:266]: Changed

**Review by thehejik — APPROVED** [review]
LGTM, thanks

**Pull request merged**